# 对数几率回归

## 实验内容
1. 使用对数几率回归完成Breast_Cancer_Wisconsin结果预测问题。
2. 计算十折交叉验证下的精度(accuracy)，查准率(precision)，查全率(recall)，F1值。

## 评测指标  
1. 精度
2. 查准率
3. 查全率
4. F1

# 1. 读取数据

In [39]:
import numpy as np
import pandas as pd

In [40]:
data = pd.read_csv('data\Breast_Cancer_Wisconsin\data.csv')

# 2. 导入模型

In [41]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
from sklearn.model_selection import cross_val_predict
from sklearn import preprocessing

In [42]:
data.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

数据处理

In [43]:
data = data.values 
data_x = data[:,2:-1]
data_y = data[:,1:2]

In [44]:
data_y = np.reshape(data_y,(-1))
data_y.shape

(569,)

# 4. 训练并预测

请使用预处理后的数据进行模型的训练与预测

In [45]:
# YOUR CODE HERE
# M为恶性，B为良性
data_y = np.where(data_y == 'M', 1, 0) 
# 特征标准化
scaler = preprocessing.StandardScaler().fit(data_x)
data_x_std = scaler.transform(data_x)
# 创建逻辑回归模型并进行十折交叉验证
model = LogisticRegression(random_state=0)
y_pred = cross_val_predict(model, data_x_std, data_y, cv=10)

# 5. 评价指标的计算

请你计算模型的四项指标

In [46]:
# YOUR CODE HERE
accuracy = accuracy_score(data_y, y_pred)
precision = precision_score(data_y, y_pred)
recall = recall_score(data_y, y_pred)
f1 = f1_score(data_y, y_pred)
print("Accuracy: ",accuracy,"\nPrecision: ",precision,"\nRecall: ",recall,"\nF1: ",f1)

Accuracy:  0.9806678383128296 
Precision:  0.9855072463768116 
Recall:  0.9622641509433962 
F1:  0.973747016706444


###### 双击此处填写

数据集|精度|查准率|查全率|F1
-|-|-|-|-
Breast_Cancer_Wisconsin | 0.9806678383128296 | 0.9855072463768116 | 0.9622641509433962 | 0.973747016706444

# 6. 选做：尝试对特征进行变换、筛选、组合后，训练模型并计算十折交叉验证后的四项指标

In [9]:
# YOUR CODE HERE



###### 双击此处填写
1. 模型1的处理流程：
2. 模型2的处理流程：
3. 模型3的处理流程:

模型|数据集|精度|查准率|查全率|F1
-|-|-|-|-|-
模型1 | 数据集 | 0.0 | 0.0 | 0.0 | 0.0
模型2 | 数据集 | 0.0 | 0.0 | 0.0 | 0.0
模型3 | 数据集 | 0.0 | 0.0 | 0.0 | 0.0